paccmann_generator/reinforce_sets.py

```diff
@@ -0,0 +1,485 @@
+        )
+
+        pred, pred_dict = self.efficacy_predictor(
+            smiles_tensor, gep_t.repeat(len(valid_smiles), 1)
+        )
+        log_micromolar_pred = self.get_log_molar(np.squeeze(pred.detach().numpy()))
+        lmps = [lmp if lmp < self.ic50_threshold else 10 for lmp in log_micromolar_pred]
+        return 1 / (1 + np.exp(lmps))
+
+    def policy_gradient(
+        self, cell_line: str, protein: str, epoch: int, batch_size: int = 10
+    ) -> Tuple:
+        """Implementation of the policy gradient algorithm.
+        
+        Args:
+            protein (str): Name of protein to generate drugs.
+            cell_line (str): Name of cell line to generate drugs.
+            epoch (int): training epoch.
+            batch_size (int, optional): Batch size. Defaults to 10.
+
+        Returns:
+            Tuple: Tuple containing the total reward and total loss.
+        """
+        rl_loss = 0
+        self.optimizer.zero_grad()
+
+        # get set
+        input_set, gep_t, protein_predictor = self.get_set(
+            cell_line, protein, batch_size, shuffle=True
+        )
+        # Encode the set
+        c_n, h_n, r_n = self.encoder(input_set)
+        latent_z = torch.mean(torch.stack([c_n, h_n, r_n]), dim=0)
+        # Produce molecules
+        valid_smiles, valid_nums, valid_idx = self.get_smiles_from_latent(latent_z)
+
+        # Get rewards (list, one reward for each valid smiles)
+        rewards = self.get_reward(valid_smiles, protein, cell_line)
+        reward_tensor = torch.unsqueeze(torch.Tensor(rewards), 1)
+
+        padded_nums = torch.nn.utils.rnn.pad_sequence(valid_nums)
+        num_mols = padded_nums.shape[1]
+
+        self.generator.decoder._update_batch_size(num_mols)
+
+        # Batch processing
+        lrps = 1
+        if self.generator.decoder.latent_dim == 2 * self.encoder.latent_size:
+            lrps = 2
+        hidden = self.generator.decoder.latent_to_hidden(
+            latent_z.repeat(
+                self.generator.decoder.n_layers, 1, lrps
+            )[:, valid_idx, :]
+        )  # yapf: disable
+        stack = self.generator.decoder.init_stack
+
+        # Compute loss
+        output = torch.Tensor()
+        for p in range(len(padded_nums) - 1):
+            output, hidden, stack = self.generator.decoder(
+                torch.unsqueeze(padded_nums[p], 0), hidden, stack
+            )
+            output = self.generator.decoder.output_layer(output).squeeze()
+
+            log_probs = F.log_softmax(output, dim=1)
+            target_char = torch.unsqueeze(padded_nums[p + 1], 1)
+            rl_loss -= torch.mean(log_probs.gather(1, target_char) * reward_tensor)
+
+        summed_reward = torch.mean(torch.Tensor(rewards))
+        rl_loss.backward()
+
+        if self.grad_clipping is not None:
+            torch.nn.utils.clip_grad_norm_(
+                list(self.generator.decoder.parameters()) +
+                list(self.encoder.parameters()), self.grad_clipping
```

Comment:
the optimizer of the base class ignores the encoder.parameters and only optimizes the decoder. Hence this is useless and might even cause errors.